// @surgent/pay-convex/dist/index.js
/**
 * Surpay Convex Integration
 *
 * Provides Convex actions that wrap the Surpay SDK with auth context support.
 *
 * @example
 * ```typescript
 * // convex/surpay.ts
 * import { Surpay } from "@surgent/pay-convex";
 *
 * const surpay = new Surpay({
 *   apiKey: process.env.SURGENT_API_KEY!,
 *   identify: async (ctx) => {
 *     const identity = await ctx.auth.getUserIdentity();
 *     if (!identity) return null;
 *     return {
 *       customerId: identity.subject,
 *       customerData: { name: identity.name, email: identity.email },
 *     };
 *   },
 * });
 *
 * export const {
 *   createCheckout,
 *   guestCheckout,
 *   check,
 *   listProducts,
 *   getCustomer,
 *   listCustomers,
 *   listSubscriptions,
 * } = surpay.api();
 * ```
 */
import { actionGeneric } from "convex/server";
import { Surpay as SurpayClient } from "@surgent/pay";
import { CreateCheckoutArgs, CheckArgs, GetCustomerArgs, ListCustomersArgs, ListSubscriptionsArgs, ListProductsArgs, GuestCheckoutArgs, } from "./types.js";
function toPlainError(error) {
    if (error instanceof Error) {
        return { message: error.message, code: error.code };
    }
    return { message: String(error) };
}
// Wrap SDK calls to ensure errors are plain objects
async function wrapSdkCall(fn) {
    try {
        const result = await fn();
        if (result.error) {
            return { data: null, error: toPlainError(result.error) };
        }
        return result;
    }
    catch (e) {
        return { data: null, error: toPlainError(e) };
    }
}
// ============================================================================
// Helpers
// ============================================================================
function resolveProductId(args, products) {
    if (args.productId)
        return args.productId;
    if (args.productSlug) {
        const found = products.find((p) => p.product.slug === args.productSlug);
        if (!found)
            throw new Error(`Product with slug "${args.productSlug}" not found`);
        return found.product.id;
    }
    throw new Error("Either productId or productSlug is required");
}
// ============================================================================
// Main Class
// ============================================================================
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class Surpay {
    client;
    options;
    constructor(config) {
        this.options = config;
        const g = globalThis;
        const envBaseUrl = g.process?.env.SURPAY_BASE_URL;
        this.client = new SurpayClient({
            apiKey: config.apiKey,
            baseUrl: config.baseUrl ?? envBaseUrl,
            responseCase: config.responseCase ?? "camel",
        });
    }
    async getIdentifierOpts(ctx) {
        return await this.options.identify(ctx);
    }
    async requireAuth(ctx) {
        const opts = await this.getIdentifierOpts(ctx);
        if (!opts) {
            throw new Error("Authentication required. Make sure the user is signed in and your identify() function returns their customerId.");
        }
        return opts;
    }
    api() {
        const client = this.client;
        return {
            /**
             * Create a checkout session for the authenticated user.
             * Requires user to be signed in (uses identify() for customerId).
             */
            createCheckout: actionGeneric({
                args: CreateCheckoutArgs,
                handler: async (ctx, args) => {
                    try {
                        const identifierOpts = await this.requireAuth(ctx);
                        // Resolve productId from slug if needed
                        let productId = args.productId;
                        if (!productId && args.productSlug) {
                            const { data: products, error } = await client.products.listWithPrices();
                            if (error)
                                return { data: null, error: toPlainError(error) };
                            productId = resolveProductId(args, products);
                        }
                        if (!productId) {
                            return { data: null, error: { message: "Either productId or productSlug is required" } };
                        }
                        return wrapSdkCall(() => client.checkout.create({
                            productId,
                            priceId: args.priceId,
                            successUrl: args.successUrl,
                            customerId: identifierOpts.customerId,
                            customerEmail: identifierOpts.customerData?.email,
                            customerName: identifierOpts.customerData?.name,
                        }));
                    }
                    catch (e) {
                        return { data: null, error: toPlainError(e) };
                    }
                },
            }),
            /**
             * Create a checkout session for a guest (anonymous) user.
             * Does NOT require authentication - customerId must be provided explicitly.
             * Use this for anonymous checkout flows.
             */
            guestCheckout: actionGeneric({
                args: GuestCheckoutArgs,
                handler: async (_ctx, args) => {
                    try {
                        // Resolve productId from slug if needed
                        let productId = args.productId;
                        if (!productId && args.productSlug) {
                            const { data: products, error } = await client.products.listWithPrices();
                            if (error)
                                return { data: null, error: toPlainError(error) };
                            productId = resolveProductId(args, products);
                        }
                        if (!productId) {
                            return { data: null, error: { message: "Either productId or productSlug is required" } };
                        }
                        return wrapSdkCall(() => client.checkout.create({
                            productId,
                            priceId: args.priceId,
                            successUrl: args.successUrl,
                            customerId: args.customerId,
                            customerEmail: args.customerEmail,
                            customerName: args.customerName,
                        }));
                    }
                    catch (e) {
                        return { data: null, error: toPlainError(e) };
                    }
                },
            }),
            /**
             * Check if a user has access to a product.
             * Pass `customerId` explicitly for guest flows, or omit it to resolve from identity.
             */
            check: actionGeneric({
                args: CheckArgs,
                handler: async (ctx, args) => {
                    try {
                        const customerId = args.customerId ?? (await this.requireAuth(ctx)).customerId;
                        // Resolve productId from slug if needed
                        let productId = args.productId;
                        if (!productId && args.productSlug) {
                            const { data: products, error } = await client.products.listWithPrices();
                            if (error)
                                return { data: null, error: toPlainError(error) };
                            productId = resolveProductId(args, products);
                        }
                        if (!productId) {
                            return { data: null, error: { message: "Either productId or productSlug is required" } };
                        }
                        return wrapSdkCall(() => client.check({
                            productId,
                            customerId,
                        }));
                    }
                    catch (e) {
                        return { data: null, error: toPlainError(e) };
                    }
                },
            }),
            /**
             * List all products with their prices.
             * Does not require authentication.
             */
            listProducts: actionGeneric({
                args: ListProductsArgs,
                handler: async () => {
                    return wrapSdkCall(() => client.products.listWithPrices());
                },
            }),
            /**
             * Get a specific customer by ID.
             * Does not require the caller to be that customer.
             */
            getCustomer: actionGeneric({
                args: GetCustomerArgs,
                handler: async (_ctx, args) => {
                    return wrapSdkCall(() => client.customers.get(args.customerId));
                },
            }),
            /**
             * List all customers (admin operation).
             */
            listCustomers: actionGeneric({
                args: ListCustomersArgs,
                handler: async () => {
                    return wrapSdkCall(() => client.customers.list());
                },
            }),
            /**
             * List all subscriptions (admin operation).
             */
            listSubscriptions: actionGeneric({
                args: ListSubscriptionsArgs,
                handler: async () => {
                    return wrapSdkCall(() => client.subscriptions.list());
                },
            }),
        };
    }
}
// ============================================================================
// Exports
// ============================================================================
export * from "./types.js";
