// @surgent/pay/dist/client.js
/**
 * Base HTTP client for Surpay SDK.
 *
 * Provides core HTTP methods with Result pattern error handling.
 */
import { toResult } from './utils/result.js';
import { SurpayError } from './errors.js';
const DEFAULT_BASE_URL = 'https://pay.surgent.dev';
const DEFAULT_TIMEOUT_MS = 30000;
export class SurpayClient {
    baseUrl;
    headers;
    resultOptions;
    timeoutMs;
    constructor(options) {
        this.baseUrl = options.baseUrl || DEFAULT_BASE_URL;
        this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
        this.headers = {
            Authorization: `Bearer ${options.apiKey}`,
            'Content-Type': 'application/json',
        };
        this.resultOptions = {
            responseCase: options.responseCase ?? 'camel',
        };
    }
    async get(path) {
        return this.request('GET', path);
    }
    async post(path, body) {
        return this.request('POST', path, body);
    }
    async put(path, body) {
        return this.request('PUT', path, body);
    }
    async delete(path) {
        return this.request('DELETE', path);
    }
    async request(method, path, body) {
        try {
            const response = await fetch(`${this.baseUrl}${path}`, {
                method,
                headers: this.headers,
                body: body !== undefined ? JSON.stringify(body) : undefined,
                signal: AbortSignal.timeout(this.timeoutMs),
            });
            return toResult(response, this.resultOptions);
        }
        catch (error) {
            return this.handleFetchError(error);
        }
    }
    handleFetchError(error) {
        // Timeout errors from AbortSignal.timeout()
        if (error instanceof DOMException && error.name === 'TimeoutError') {
            return {
                data: null,
                error: new SurpayError({
                    message: `Request timed out after ${this.timeoutMs}ms`,
                    code: 'timeout_error',
                    statusCode: 0,
                }),
                statusCode: 0,
            };
        }
        // Network errors (DNS failure, connection refused, etc.)
        const message = error instanceof Error ? error.message : 'Network request failed';
        return {
            data: null,
            error: new SurpayError({
                message,
                code: 'network_error',
                statusCode: 0,
            }),
            statusCode: 0,
        };
    }
}
