// @surgent/pay/dist/surpay.js
/**
 * Surpay Client
 *
 * Main SDK client for tenant-level operations.
 * Uses grouped method namespaces (Autumn pattern).
 */
import { SurpayClient } from './client.js';
export class Surpay extends SurpayClient {
    constructor(options) {
        const envApiKey = typeof process !== 'undefined' ? process.env.SURPAY_API_KEY : undefined;
        const envBaseUrl = typeof process !== 'undefined' ? process.env.SURPAY_BASE_URL : undefined;
        const apiKey = options?.apiKey || envApiKey || '';
        const baseUrl = options?.baseUrl ?? envBaseUrl;
        if (!apiKey) {
            throw new Error('Surpay API key is required. Pass it via options or set SURPAY_API_KEY env var.');
        }
        super({ apiKey, baseUrl, responseCase: options?.responseCase });
    }
    customers = {
        list: () => this.get('/customers'),
        get: (customerId) => this.get(`/customers/${customerId}`),
    };
    products = {
        create: (params) => this.post('/product', params),
        update: (productId, params) => this.put(`/product/${productId}`, params),
        listWithPrices: () => this.get('/products'),
    };
    prices = {
        create: (params) => this.post('/product/price', params),
    };
    checkout = {
        create: (params) => this.post('/checkout', params),
    };
    check = (params) => this.post('/check', params);
    subscriptions = {
        list: () => this.get('/subscriptions'),
    };
    transactions = {
        list: () => this.get('/transactions'),
    };
    projects = {
        list: async () => {
            const result = await this.get('/projects');
            if (result.error) {
                return result;
            }
            // Normalize wrapped response: { projects: [...] } -> [...]
            const data = Array.isArray(result.data) ? result.data : result.data.projects || [];
            return { data, error: null, statusCode: result.statusCode };
        },
    };
    accounts = {
        connect: (params) => this.post('/accounts/connect/whop', params),
        get: (accountId) => this.get(`/accounts/${accountId}`),
        list: () => this.get('/accounts'),
        delete: (accountId) => this.delete(`/accounts/${accountId}`),
    };
}
